reflow/dtls_wrapper/bf_dwrap.c

```diff
@@ -21,50 +56,60 @@ static int dwrap_gets(BIO *b, char *buf, int size);
 static long dwrap_ctrl(BIO *b, int cmd, long num, void *ptr);
 static long dwrap_callback_ctrl(BIO *b, int cmd, bio_info_cb *fp);
 
-static BIO_METHOD methods_dwrap=
-{
-   BIO_TYPE_DWRAP,
-      "dtls_wrapper",
-      dwrap_write,
-      dwrap_read,
-      dwrap_puts,
-      dwrap_gets,
-      dwrap_ctrl,
-      dwrap_new,
-      dwrap_free,
-      dwrap_callback_ctrl
-};
-
 typedef struct BIO_F_DWRAP_CTX_ 
 {
    int dgram_timer_exp;
 } BIO_F_DWRAP_CTX;
 
 
-BIO_METHOD *BIO_f_dwrap(void) 
+static BIO_METHOD* dtls_meth = NULL;
+
+static void free_dtls_meth()
 {
-   return(&methods_dwrap);
+   BIO_meth_free(dtls_meth);
+   dtls_meth = NULL;
+}
+
+BIO_METHOD *BIO_f_dwrap(void)
+{
+   if (!dtls_meth) {
+      dtls_meth = BIO_meth_new(BIO_TYPE_DWRAP, "dtls_wrapper");
```

Comment:
This is not thread-safe. Not sure if there are cases when this code can run in different threads, though. I would recommend using `pthread_once` or atomics instead.

Reply:
You're right but this codebase is still C++03. How about using a static `auto_ptr` holding the `BIO_method`? That would also enable me to get rid of the `atexit` hack.

Reply:
A simpler solution is to convert `dtls_meth` to an object with a constructor and destructor and move the `BIO_METHOD` initialization to the constructor. This would still be not thread-safe, but in practical sense this would be better because the constructor will be run at library load time, when there are typically no multiple threads.

Reply:
OTOH who knows if calling into OpenSSL works before initalization ...

Reply:
> How about using a static auto_ptr holding the BIO_method?

This won't help. The race is between loading the pointer for testing against `NULL` and storing the allocated pointer. `auto_ptr` doesn't help with it.

Re atomics, I was meaning compiler-specific intrinsics. One would likely have to introduce some abstraction layer for portability. `pthread_once` is available for C++03 programs but not on Windows. I would probably use `Interlocked*` intrinsics on Windows and `pthread_once` everywhere else.

Reply:
Will work on something like this. But I have to finish some unrelated work first.

Reply:
> OTOH who knows if calling into OpenSSL works before initalization

I would assume it should work for the methods we need - those are just memory allocation and accessors, no apparent reason for them to not work.

Reply:
@Lastique I changed the code as you suggested. Could you pleas give it another review?
 
Thanks a lot for your time!